File: bwl.Imaging/RawFrames/RawIntFrameConverters.vb
Imports System.Runtime.CompilerServices
Imports System.Threading.Tasks
Imports Bwl.Imaging

Public Module RawIntFrameConverters

    <Extension()>
    Public Function ConvertTo8Bit(frame As RawIntFrame, gainByBitOffset As Integer) As RGBMatrix
        Dim mtr1 As New RGBMatrix(frame.Width, frame.Height)
        For y = 0 To frame.Height - 1
            For x = 0 To frame.Width - 1
                Dim b = frame.Data((x + y * frame.Width) * 3 + 0)
                Dim g = frame.Data((x + y * frame.Width) * 3 + 1)
                Dim r = frame.Data((x + y * frame.Width) * 3 + 2)
                If gainByBitOffset <= 4 Then
                    r = r >> (4 - gainByBitOffset)
                    g = g >> (4 - gainByBitOffset)
                    b = b >> (4 - gainByBitOffset)
                Else
                    r = r << (gainByBitOffset - 4)
                    g = g << (gainByBitOffset - 4)
                    b = b << (gainByBitOffset - 4)
                End If
                mtr1.RedPixel(x, y) = Math.Min(r, 255)
                mtr1.GreenPixel(x, y) = Math.Min(g, 255)
                mtr1.BluePixel(x, y) = Math.Min(b, 255)
            Next
        Next
        Return mtr1
    End Function

    <Extension()>
    Public Function ConvertTo8BitFast(frame As RawIntFrame, gainByBitOffset As Integer) As RGBMatrix
        Dim mtr1 As New RGBMatrix(frame.Width, frame.Height)
        Dim frameData = frame.Data
        If gainByBitOffset <= 4 Then
            Parallel.For(0, 3, Sub(channel As Integer)
                                   Dim matrix = mtr1.Matrix(2 - channel)
                                   For i = 0 To matrix.Length - 1
                                       Dim px = frameData(i * 3 + channel) >> (4 - gainByBitOffset)
                                       matrix(i) = Math.Min(px, 255)
                                   Next
                               End Sub)
        Else
            Parallel.For(0, 3, Sub(channel As Integer)
                                   Dim matrix = mtr1.Matrix(2 - channel)
                                   For i = 0 To matrix.Length - 1
                                       Dim px = frameData(i * 3 + channel) << (gainByBitOffset - 4)
                                       matrix(i) = Math.Min(px, 255)
                                   Next
                               End Sub)
        End If
        Return mtr1
    End Function

    Public Function ConvertTo8BitPair(frame As RawIntFrame, Optional totalBits As Integer = 12) As RGBMatrix()
        Dim mtr1 As New RGBMatrix(frame.Width, frame.Height)
        Dim mtr2 As New RGBMatrix(frame.Width, frame.Height)

        For y = 0 To frame.Height - 1
            For x = 0 To frame.Width - 1
                Dim b = frame.Data((x + y * frame.Width) * 3 + 0)
                Dim g = frame.Data((x + y * frame.Width) * 3 + 1)
                Dim r = frame.Data((x + y * frame.Width) * 3 + 2)

                If totalBits = 12 Then
                    Dim rh = (r >> 4)
                    Dim rl = (r)

                    Dim bh = (b >> 4)
                    Dim bl = (b)

                    Dim gh = (g >> 4)
                    Dim gl = (g)

                    If rl > 63 Then rl = 63
                    If gl > 63 Then gl = 63
                    If bl > 63 Then bl = 63

                    'If rl > 31 Then rl = rl And 15 Or 16
                    'If gl > 31 Then gl = gl And 15 Or 16
                    'If bl > 31 Then bl = bl And 15 Or 16

                    mtr1.RedPixel(x, y) = rh
                    mtr1.GreenPixel(x, y) = gh
                    mtr1.BluePixel(x, y) = bh

                    mtr2.RedPixel(x, y) = rl
                    mtr2.GreenPixel(x, y) = gl
                    mtr2.BluePixel(x, y) = bl
                End If
            Next
        Next
        Return {mtr1, mtr2}
    End Function

    Public Function ConvertFrom8BitPair(pair As RGBMatrix(), Optional totalBits As Integer = 12) As RawIntFrame
        Dim mtr1 = pair(0)
        Dim mtr2 = pair(1)
        Dim arr(mtr1.Width * mtr1.Height * 3 - 1) As Integer
        Dim frame As New RawIntFrame(mtr1.Width, mtr2.Height, arr)

        For y = 0 To mtr1.Height - 1
            For x = 0 To mtr1.Width - 1

                Dim rh = mtr1.RedPixel(x, y)
                Dim gh = mtr1.GreenPixel(x, y)
                Dim bh = mtr1.BluePixel(x, y)

                Dim rl = mtr2.RedPixel(x, y)
                Dim gl = mtr2.GreenPixel(x, y)
                Dim bl = mtr2.BluePixel(x, y)

                If totalBits = 12 Then
                    Dim r = ((rh And &HFF) << 4) Or ((rl And &HF))
                    Dim g = ((gh And &HFF) << 4) Or ((gl And &HF))
                    Dim b = ((bh And &HFF) << 4) Or ((bl And &HF))

                    frame.Data((x + y * frame.Width) * 3 + 0) = b
                    frame.Data((x + y * frame.Width) * 3 + 1) = g
                    frame.Data((x + y * frame.Width) * 3 + 2) = r
                End If
            Next
        Next
        Return frame
    End Function

    <Extension()>
    Public Function ConvertHDR1(frame As RawIntFrame, baseGain As Integer) As RGBMatrix
        Dim mtr1 As New RGBMatrix(frame.Width, frame.Height)
        For i = 0 To frame.Width * frame.Height - 1
            Dim b = frame.Data(i * 3 + 0)
            Dim g = frame.Data(i * 3 + 1)
            Dim r = frame.Data(i * 3 + 2)

            If baseGain <= 4 Then
                r = r >> (4 - baseGain)
                g = g >> (4 - baseGain)
                b = b >> (4 - baseGain)
            Else
                r = r << (baseGain - 4)
                g = g << (baseGain - 4)
                b = b << (baseGain - 4)
            End If
            Dim p = 255
            Dim t = 60
            Dim k = 0.8
            Do While (r > p Or g > p Or b > p) 'And (Math.Abs(r - b) > t Or Math.Abs(g - b) > t Or Math.Abs(g - r) > t)
                '    Do While (r > p Or g > p Or b > p) And (Math.Abs(r - b) > t Or Math.Abs(g - b) > t Or Math.Abs(g - r) > t)
                '    Do While (r > p Or g > p Or b > p) And Not (r > t And g > t And b > t)
                r = r * k
                g = g * k
                b = b * k
            Loop
            mtr1.Red(i) = Math.Min(r, 255)
            mtr1.Green(i) = Math.Min(g, 255)
            mtr1.Blue(i) = Math.Min(b, 255)
        Next
        Return mtr1
    End Function

    <Extension()>
    Public Function ConvertHDR1Fast(frame As RawIntFrame, baseGain As Integer) As RGBMatrix
        Dim mtr1 As New RGBMatrix(frame.Width, frame.Height)
        Dim frameData = frame.Data
        Dim p = 255
        Dim k = 14
        If baseGain <= 4 Then
            Dim bitShift = 4 - baseGain
            Parallel.For(0, frame.Width * frame.Height, Sub(i As Integer)
                                                            Dim b = frameData(i * 3 + 0) >> bitShift
                                                            Dim g = frameData(i * 3 + 1) >> bitShift
                                                            Dim r = frameData(i * 3 + 2) >> bitShift
                                                            While (r > p Or g > p Or b > p)
                                                                r = (r * k) >> 4
                                                                g = (g * k) >> 4
                                                                b = (b * k) >> 4
                                                            End While
                                                            mtr1.Red(i) = r
                                                            mtr1.Green(i) = g
                                                            mtr1.Blue(i) = b
                                                        End Sub)
        Else
            Dim bitShift = baseGain - 4
            Parallel.For(0, frame.Width * frame.Height, Sub(i As Integer)
                                                            Dim b = frameData(i * 3 + 0) << bitShift
                                                            Dim g = frameData(i * 3 + 1) << bitShift
                                                            Dim r = frameData(i * 3 + 2) << bitShift
                                                            While (r > p Or g > p Or b > p)
                                                                r = (r * k) >> 4
                                                                g = (g * k) >> 4
                                                                b = (b * k) >> 4
                                                            End While
                                                            mtr1.Red(i) = r
                                                            mtr1.Green(i) = g
                                                            mtr1.Blue(i) = b
                                                        End Sub)
        End If
        Return mtr1
    End Function

    <Extension()>
    Public Function ConvertHDR2(frame As RawIntFrame, baseGain As Integer) As RGBMatrix
        Dim mtr1 As New RGBMatrix(frame.Width, frame.Height)
        Dim max1 As New GrayMatrix(frame.Width / 4 + 8, frame.Height / 4 + 8)
        Dim min1 As New GrayMatrix(frame.Width / 4 + 8, frame.Height / 4 + 8)

        For y = 0 To frame.Height - 1
            For x = 0 To frame.Width - 1
                Dim b = frame.Data((x + y * frame.Width) * 3 + 0)
                Dim g = frame.Data((x + y * frame.Width) * 3 + 1)
                Dim r = frame.Data((x + y * frame.Width) * 3 + 2)

                If baseGain > 4 Then baseGain = 4
                '     r = r >> (4 - baseGain)
                '    g = g >> (4 - baseGain)
                '    b = b >> (4 - baseGain)

                Dim max = Math.Max(Math.Max(r, g), b)
                Dim min = Math.Min(Math.Min(r, g), b)
                max1.GrayPixel(x / 4, y / 4) = max / 16
                min1.GrayPixel(x / 4, y / 4) = min / 16

                mtr1.RedPixel(x, y) = r
                mtr1.GreenPixel(x, y) = g
                mtr1.BluePixel(x, y) = b
            Next
        Next
        Dim bmp = max1.ToBitmap

        Dim n = 4
        Dim k = 256
        max1 = Filters.MedianFilter2D(max1, n)
        For y = 0 To frame.Height - 1
            For x = 0 To frame.Width - 1
                Dim b = frame.Data((x + y * frame.Width) * 3 + 0)
                Dim g = frame.Data((x + y * frame.Width) * 3 + 1)
                Dim r = frame.Data((x + y * frame.Width) * 3 + 2)

                Dim min = min1.GrayPixel(x / 4, y / 4) * 16
                Dim max = max1.GrayPixel(x / 4, y / 4) * 16

                If max > 255 Then
                    r = r * k / max
                    g = g * k / max
                    b = b * k / max
                End If

                mtr1.RedPixel(x, y) = Math.Min(r, 255)
                mtr1.GreenPixel(x, y) = Math.Min(g, 255)
                mtr1.BluePixel(x, y) = Math.Min(b, 255)
            Next
        Next
        Return mtr1
    End Function

    <Extension()>
    Public Function ConvertHDR3(frame As RawIntFrame) As RGBMatrix
        Dim mtr1 As New RGBMatrix(frame.Width, frame.Height)
        For y = 0 To frame.Height - 1
            For x = 0 To frame.Width - 1
                Dim b = frame.Data((x + y * frame.Width) * 3 + 0)
                Dim g = frame.Data((x + y * frame.Width) * 3 + 1)
                Dim r = frame.Data((x + y * frame.Width) * 3 + 2)

                Dim v = 0.4
                Dim k = 12
                Dim m = 0
                If r < m Then r = m
                If g < m Then g = m
                If b < m Then b = m
                b = Math.Pow((b - m), v) * k
                g = Math.Pow((g - m), v) * k
                r = Math.Pow((r - m), v) * k

                mtr1.RedPixel(x, y) = Math.Min(r, 255)
                mtr1.GreenPixel(x, y) = Math.Min(g, 255)
                mtr1.BluePixel(x, y) = Math.Min(b, 255)
            Next
        Next
        Return mtr1
    End Function

    <Extension()>
    Public Function ConvertHDR3Fast(frame As RawIntFrame) As RGBMatrix
        Static powTable As Integer()
        If powTable Is Nothing Then
            powTable = New Integer((1 << 12) - 1) {}
            Dim v = 0.4
            Dim k = 12
            For powArg = 0 To powTable.Length - 1
                powTable(powArg) = Math.Min(Math.Pow(powArg, v) * k, 255)
            Next
        End If
        Dim mtr1 As New RGBMatrix(frame.Width, frame.Height)
        Dim frameData = frame.Data
        Parallel.For(0, 3, Sub(channel As Integer)
                               Dim matrix = mtr1.Matrix(2 - channel)
                               For i = 0 To matrix.Length - 1
                                   Dim px = frameData(i * 3 + channel)
                                   matrix(i) = powTable(px)
                               Next
                           End Sub)
        Return mtr1
    End Function
End Module

File: Bwl.Imaging.Skia/RawFrames/Integer/RawIntFrameConverters.cs
using System;
using System.Threading.Tasks;

namespace Bwl.Imaging.Skia
{

    public static class RawIntFrameConverters
    {

        public static RGBMatrix ConvertTo8Bit(this RawIntFrame frame, int gainByBitOffset)
        {
            var mtr1 = new RGBMatrix(frame.Width, frame.Height);
            for (int y = 0, loopTo = frame.Height - 1; y <= loopTo; y++)
            {
                for (int x = 0, loopTo1 = frame.Width - 1; x <= loopTo1; x++)
                {
                    int b = frame.Data[(x + y * frame.Width) * 3 + 0];
                    int g = frame.Data[(x + y * frame.Width) * 3 + 1];
                    int r = frame.Data[(x + y * frame.Width) * 3 + 2];
                    if (gainByBitOffset <= 4)
                    {
                        r = r >> 4 - gainByBitOffset;
                        g = g >> 4 - gainByBitOffset;
                        b = b >> 4 - gainByBitOffset;
                    }
                    else
                    {
                        r = r << gainByBitOffset - 4;
                        g = g << gainByBitOffset - 4;
                        b = b << gainByBitOffset - 4;
                    }
                    mtr1.SetRedPixel(x, y, Math.Min(r, 255));
                    mtr1.SetGreenPixel(x, y, Math.Min(g, 255));
                    mtr1.SetBluePixel(x, y, Math.Min(b, 255));
                }
            }
            return mtr1;
        }

        public static RGBMatrix ConvertTo8BitFast(this RawIntFrame frame, int gainByBitOffset)
        {
            var mtr1 = new RGBMatrix(frame.Width, frame.Height);
            int[] frameData = frame.Data;
            if (gainByBitOffset <= 4)
            {
                Parallel.For(0, 3, (channel) =>
                    {
                        int[] matrix = mtr1.GetMatrix(2 - channel);
                        for (int i = 0, loopTo = matrix.Length - 1; i <= loopTo; i++)
                        {
                            int px = frameData[i * 3 + channel] >> 4 - gainByBitOffset;
                            matrix[i] = Math.Min(px, 255);
                        }
                    });
            }
            else
            {
                Parallel.For(0, 3, (channel) =>
                    {
                        int[] matrix = mtr1.GetMatrix(2 - channel);
                        for (int i = 0, loopTo = matrix.Length - 1; i <= loopTo; i++)
                        {
                            int px = frameData[i * 3 + channel] << gainByBitOffset - 4;
                            matrix[i] = Math.Min(px, 255);
                        }
                    });
            }
            return mtr1;
        }

        public static RGBMatrix[] ConvertTo8BitPair(RawIntFrame frame, int totalBits = 12)
        {
            var mtr1 = new RGBMatrix(frame.Width, frame.Height);
            var mtr2 = new RGBMatrix(frame.Width, frame.Height);

            for (int y = 0, loopTo = frame.Height - 1; y <= loopTo; y++)
            {
                for (int x = 0, loopTo1 = frame.Width - 1; x <= loopTo1; x++)
                {
                    int b = frame.Data[(x + y * frame.Width) * 3 + 0];
                    int g = frame.Data[(x + y * frame.Width) * 3 + 1];
                    int r = frame.Data[(x + y * frame.Width) * 3 + 2];

                    if (totalBits == 12)
                    {
                        int rh = r >> 4;
                        int rl = r;

                        int bh = b >> 4;
                        int bl = b;

                        int gh = g >> 4;
                        int gl = g;

                        if (rl > 63)
                            rl = 63;
                        if (gl > 63)
                            gl = 63;
                        if (bl > 63)
                            bl = 63;

                        // If rl > 31 Then rl = rl And 15 Or 16
                        // If gl > 31 Then gl = gl And 15 Or 16
                        // If bl > 31 Then bl = bl And 15 Or 16

                        mtr1.SetRedPixel(x, y, rh);
                        mtr1.SetGreenPixel(x, y, gh);
                        mtr1.SetBluePixel(x, y, bh);

                        mtr2.SetRedPixel(x, y, rl);
                        mtr2.SetGreenPixel(x, y, gl);
                        mtr2.SetBluePixel(x, y, bl);
                    }
                }
            }
            return new[] { mtr1, mtr2 };
        }

        public static RawIntFrame ConvertFrom8BitPair(RGBMatrix[] pair, int totalBits = 12)
        {
            var mtr1 = pair[0];
            var mtr2 = pair[1];
            var arr = new int[(mtr1.Width * mtr1.Height * 3)];
            var frame = new RawIntFrame(mtr1.Width, mtr2.Height, arr);

            for (int y = 0, loopTo = mtr1.Height - 1; y <= loopTo; y++)
            {
                for (int x = 0, loopTo1 = mtr1.Width - 1; x <= loopTo1; x++)
                {

                    int rh = mtr1.GetRedPixel(x, y);
                    int gh = mtr1.GetGreenPixel(x, y);
                    int bh = mtr1.GetBluePixel(x, y);

                    int rl = mtr2.GetRedPixel(x, y);
                    int gl = mtr2.GetGreenPixel(x, y);
                    int bl = mtr2.GetBluePixel(x, y);

                    if (totalBits == 12)
                    {
                        int r = (rh & 0xFF) << 4 | rl & 0xF;
                        int g = (gh & 0xFF) << 4 | gl & 0xF;
                        int b = (bh & 0xFF) << 4 | bl & 0xF;

                        frame.Data[(x + y * frame.Width) * 3 + 0] = b;
                        frame.Data[(x + y * frame.Width) * 3 + 1] = g;
                        frame.Data[(x + y * frame.Width) * 3 + 2] = r;
                    }
                }
            }
            return frame;
        }

        public static RGBMatrix ConvertHDR1(this RawIntFrame frame, int baseGain)
        {
            var mtr1 = new RGBMatrix(frame.Width, frame.Height);
            for (int i = 0, loopTo = frame.Width * frame.Height - 1; i <= loopTo; i++)
            {
                int b = frame.Data[i * 3 + 0];
                int g = frame.Data[i * 3 + 1];
                int r = frame.Data[i * 3 + 2];

                if (baseGain <= 4)
                {
                    r = r >> 4 - baseGain;
                    g = g >> 4 - baseGain;
                    b = b >> 4 - baseGain;
                }
                else
                {
                    r = r << baseGain - 4;
                    g = g << baseGain - 4;
                    b = b << baseGain - 4;
                }
                int p = 255;
                int t = 60;
                double k = 0.8d;
                while (r > p | g > p | b > p) // And (Math.Abs(r - b) > t Or Math.Abs(g - b) > t Or Math.Abs(g - r) > t)
                {
                    // Do While (r > p Or g > p Or b > p) And (Math.Abs(r - b) > t Or Math.Abs(g - b) > t Or Math.Abs(g - r) > t)
                    // Do While (r > p Or g > p Or b > p) And Not (r > t And g > t And b > t)
                    r = (int)Math.Round(r * k);
                    g = (int)Math.Round(g * k);
                    b = (int)Math.Round(b * k);
                }
                mtr1.Red[i] = Math.Min(r, 255);
                mtr1.Green[i] = Math.Min(g, 255);
                mtr1.Blue[i] = Math.Min(b, 255);
            }
            return mtr1;
        }

        public static RGBMatrix ConvertHDR1Fast(this RawIntFrame frame, int baseGain)
        {
            var mtr1 = new RGBMatrix(frame.Width, frame.Height);
            int[] frameData = frame.Data;
            int p = 255;
            int k = 14;
            if (baseGain <= 4)
            {
                int bitShift = 4 - baseGain;
                Parallel.For(0, frame.Width * frame.Height, (i) =>
                    {
                        int b = frameData[i * 3 + 0] >> bitShift;
                        int g = frameData[i * 3 + 1] >> bitShift;
                        int r = frameData[i * 3 + 2] >> bitShift;
                        while (r > p | g > p | b > p)
                        {
                            r = r * k >> 4;
                            g = g * k >> 4;
                            b = b * k >> 4;
                        }
                        mtr1.Red[i] = r;
                        mtr1.Green[i] = g;
                        mtr1.Blue[i] = b;
                    });
            }
            else
            {
                int bitShift = baseGain - 4;
                Parallel.For(0, frame.Width * frame.Height, (i) =>
                    {
                        int b = frameData[i * 3 + 0] << bitShift;
                        int g = frameData[i * 3 + 1] << bitShift;
                        int r = frameData[i * 3 + 2] << bitShift;
                        while (r > p | g > p | b > p)
                        {
                            r = r * k >> 4;
                            g = g * k >> 4;
                            b = b * k >> 4;
                        }
                        mtr1.Red[i] = r;
                        mtr1.Green[i] = g;
                        mtr1.Blue[i] = b;
                    });
            }
            return mtr1;
        }

        public static RGBMatrix ConvertHDR2(this RawIntFrame frame, int baseGain)
        {
            var mtr1 = new RGBMatrix(frame.Width, frame.Height);
            var max1 = new GrayMatrix(frame.Width / 4 + 8, frame.Height / 4 + 8);
            var min1 = new GrayMatrix(frame.Width / 4 + 8, frame.Height / 4 + 8);

            for (int y = 0, loopTo = frame.Height - 1; y <= loopTo; y++)
            {
                for (int x = 0, loopTo1 = frame.Width - 1; x <= loopTo1; x++)
                {
                    int b = frame.Data[(x + y * frame.Width) * 3 + 0];
                    int g = frame.Data[(x + y * frame.Width) * 3 + 1];
                    int r = frame.Data[(x + y * frame.Width) * 3 + 2];

                    if (baseGain > 4)
                        baseGain = 4;
                    // r = r >> (4 - baseGain)
                    // g = g >> (4 - baseGain)
                    // b = b >> (4 - baseGain)

                    int max = Math.Max(Math.Max(r, g), b);
                    int min = Math.Min(Math.Min(r, g), b);
                    max1.SetGrayPixel(x / 4, y / 4, max / 16);
                    min1.SetGrayPixel(x / 4, y / 4, min / 16);

                    mtr1.SetRedPixel(x, y, r);
                    mtr1.SetGreenPixel(x, y, g);
                    mtr1.SetBluePixel(x, y, b);
                }
            }
            var bmp = max1.ToBitmap();

            int n = 4;
            int k = 256;
            max1 = Filters.MedianFilter2D(max1, n);
            for (int y = 0, loopTo2 = frame.Height - 1; y <= loopTo2; y++)
            {
                for (int x = 0, loopTo3 = frame.Width - 1; x <= loopTo3; x++)
                {
                    int b = frame.Data[(x + y * frame.Width) * 3 + 0];
                    int g = frame.Data[(x + y * frame.Width) * 3 + 1];
                    int r = frame.Data[(x + y * frame.Width) * 3 + 2];

                    int min = min1.GetGrayPixel(x / 4, y / 4) * 16;
                    int max = max1.GetGrayPixel(x / 4, y / 4) * 16;

                    if (max > 255)
                    {
                        r = r * k / max;
                        g = g * k / max;
                        b = b * k / max;
                    }

                    mtr1.SetRedPixel(x, y, Math.Min(r, 255));
                    mtr1.SetGreenPixel(x, y, Math.Min(g, 255));
                    mtr1.SetBluePixel(x, y, Math.Min(b, 255));
                }
            }
            return mtr1;
        }

        public static RGBMatrix ConvertHDR3(this RawIntFrame frame)
        {
            var mtr1 = new RGBMatrix(frame.Width, frame.Height);
            for (int y = 0, loopTo = frame.Height - 1; y <= loopTo; y++)
            {
                for (int x = 0, loopTo1 = frame.Width - 1; x <= loopTo1; x++)
                {
                    int b = frame.Data[(x + y * frame.Width) * 3 + 0];
                    int g = frame.Data[(x + y * frame.Width) * 3 + 1];
                    int r = frame.Data[(x + y * frame.Width) * 3 + 2];

                    double v = 0.4d;
                    int k = 12;
                    int m = 0;
                    if (r < m)
                        r = m;
                    if (g < m)
                        g = m;
                    if (b < m)
                        b = m;
                    b = (int)Math.Round(Math.Pow(b - m, v) * k);
                    g = (int)Math.Round(Math.Pow(g - m, v) * k);
                    r = (int)Math.Round(Math.Pow(r - m, v) * k);

                    mtr1.SetRedPixel(x, y, Math.Min(r, 255));
                    mtr1.SetGreenPixel(x, y, Math.Min(g, 255));
                    mtr1.SetBluePixel(x, y, Math.Min(b, 255));
                }
            }
            return mtr1;
        }
        // Таблицы
        private static double _ConvertHDR3Fast_vs = 0.5d;
        private static double _ConvertHDR3Fast_ks = 5d;
        private static int[] _ConvertHDR3Fast_powTable = default;

        public static RGBMatrix ConvertHDR3Fast(this RawIntFrame frame, double v = 0.5d, double k = 5d)
        {
            if (_ConvertHDR3Fast_powTable is null || _ConvertHDR3Fast_vs != v || _ConvertHDR3Fast_ks != k)
            {
                _ConvertHDR3Fast_vs = v;
                _ConvertHDR3Fast_ks = k;
                _ConvertHDR3Fast_powTable = new int[4096]; // 4095 - 12 bit
                for (int powArg = 0, loopTo = _ConvertHDR3Fast_powTable.Length - 1; powArg <= loopTo; powArg++)
                    _ConvertHDR3Fast_powTable[powArg] = (int)Math.Round(Math.Min(Math.Pow(powArg, _ConvertHDR3Fast_vs) * _ConvertHDR3Fast_ks, 255d));
            }
            // Расчет
            var mtr1 = new RGBMatrix(frame.Width, frame.Height);
            int[] frameData = frame.Data;
            Parallel.For(0, 3, (channel) =>
                {
                    int[] matrix = mtr1.GetMatrix(2 - channel);
                    for (int i = 0, loopTo = matrix.Length - 1; i <= loopTo; i++)
                    {
                        int px = frameData[i * 3 + channel];
                        matrix[i] = _ConvertHDR3Fast_powTable[px];
                    }
                });
            return mtr1;
        }
    }
}